core-blocks/paragraph/index.js

```diff
@@ -314,6 +317,8 @@ export const settings = {
 
 	attributes: schema,
 
+	controls,
```

Comment:
Last time there was some discussion for keeping the controls in the edit function: https://github.com/WordPress/gutenberg/pull/3535#issuecomment-350733235. Insure if @aduth and @mtias still feel this way.

Reply:
In passing: I still don't particularly like this being part of the block API, when otherwise we've included it (and other block controls) as part of the block's own `edit` rendering. Including it in the block API seems more of a means to an end (being multi-block control) than being strictly "correct". Of course, I don't have a great idea in mind for working around this, aside from... not doing multi-toolbars 😄 Perhaps there is something where we could fake the render once of a given consistent set of selected blocks, inspect the result to see if there are block controls, and use that.

Reply:
I tried something similar to the fake render, grabbing the controls from `<BlockControls>` and reusing them, but their `onClick` handlers were bound to functions that had the `setAttributes` and values from `this.props`, and I couldn't make them reusable. That's been what forced me into putting them into the block API - I can grab the controls from the slot, but they're bound to a single block instance.

So while I would rather have them in the edit function too, I can't see a way of reusing them for the multi block toolbar.

Reply:
FWIW, we're going to come up against the same thing for the `toolbar` too. So if there _is_ a way, we should talk about it as much as possible here so we have the right approach :)

Reply:
We've also blurred the line between how individual block units relate to one another. The original API was never meant to really support this sort of "apply behaviors to multiple instances", it was designed to support a single instance at a time. Bolting it on at this point will make for something of a Frankenstein API.

To be fair, I also lean more on the end of: If we can't make it work and have a consistent approach to block independence vs. interdependence, it shouldn't be included as a feature at all.

Reply:
I'm not sure I agree there. It's a big missing feature for end users, and shouldn't be skipped because of technical purity. Having said that, I do agree that having 90% of the API able to use `this` and the rest _not_ able to, feels quite wrong!

Perhaps we have block controls as something which decorates a block's settings, to keep the approaches distinct from each other?

From a dev perspective, something like:
```
// controls.js
export const controls = {
	toolbar: function( attributes, setAttributes ) {
		// ... do your instance agnostic toolbar controls thing there
	},
	inspector: function( attributes, setAttributes ) {
		// ... and inspector things here
	}
}


// index.js
import { withControls } from '../../blocks/api';
import controls from './controls';

const settings = withControls( controls, {
	// block settings here, using `this` happily
} );

```

Reply:
Looking more into what we can do here today. I feel like I should say, I agree that this shouldn't go in settings, and looking more at how block registration works, I don't think `withControls` fits either. Perhaps if we define `controls` as well as `settings` in a block, that could be passed to `registerBlockType`, and because it's not inside `edit`, there's no danger of trying to use `this` in a context where it shouldn't be used.

I'll put something together for comment :)